src/backend/distributed/connection/connection_management.c

```diff
@@ -437,187 +458,347 @@ ShutdownConnection(MultiConnection *connection)
+{
+	MemoryContextCallback *cb = MemoryContextAllocZero(CurrentMemoryContext,
+													   sizeof(MemoryContextCallback));
+	cb->func = callback;
+	cb->arg = arg;
+	MemoryContextRegisterResetCallback(CurrentMemoryContext, cb);
+}
+
+
+/*
+ * FinishConnectionListEstablishment is a wrapper around FinishConnectionEstablishment.
+ * The function iterates over the multiConnectionList and finishes the connection
+ * establishment for each multi connection.
+ */
+void
+FinishConnectionListEstablishment(List *multiConnectionList)
+{
+	int connectionIndex = 0;
+	int connectionCount = 0;
+	TimestampTz deadline = 0;
+	MultiConnectionReady *connections = NULL;
+	ListCell *multiConnectionCell = NULL;
+
+	WaitEventSet *waitEventSet = NULL;
+	bool waitEventSetRebuild = false;
+	int waitCount = 0;
+	WaitEvent *events = NULL;
+
+	/*
+	 * Here we prepare an array that keeps the MultiConnection together with the
+	 * information on if and how to wait before PQconnectionPoll should be called on it
+	 */
+	connectionCount = list_length(multiConnectionList);
+	connections = (MultiConnectionReady *) palloc0(connectionCount *
+												   sizeof(MultiConnectionReady));
+	foreach(multiConnectionCell, multiConnectionList)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(multiConnectionCell);
+		TimestampTz connectionDeadline = TimestampTzPlusMilliseconds(
+			connection->connectionStart, NodeConnectionTimeout);
+		if (connectionDeadline > deadline)
 		{
-			return;
+			deadline = connectionDeadline;
 		}
-		else
+
+		connections[connectionIndex].connection = connection;
+
+		/*
+		 * before we can build the waitset to wait for asynchronous IO we need to know the
+		 * pollmode to use for the sockets. This is populated by executing one round of
+		 * PQconnectPoll. This updates the MultiConnectionReady struct with the ready state
+		 * and its next poll mode.
+		 */
+		MultiConnectionReadyExecutePoll(&connections[connectionIndex]);
+
+		if (!connections[connectionIndex].ready)
 		{
-			Assert(pollmode == PGRES_POLLING_WRITING ||
-				   pollmode == PGRES_POLLING_READING);
+			waitCount++;
 		}
 
-		/* Loop, to handle poll() being interrupted by signals (EINTR) */
-		while (true)
+		connectionIndex++;
+	}
+
+	/* prepapre space for socket events */
+	events = (WaitEvent *) palloc0(connectionCount * sizeof(WaitEvent));
+
+	while (waitCount > 0)
+	{
+		long timeout = DeadlineTimestampTzToTimeout(deadline);
+		int eventCount = 0;
+		int eventIndex = 0;
+
+		if (waitEventSet == NULL || waitEventSetRebuild)
 		{
-			int pollResult = 0;
+			waitEventSet = WaitEventSetFromMultiConnectionReadyArray(connections,
+																	 connectionCount,
+																	 &waitCount);
 
-			/* we use poll(2) if available, otherwise select(2) */
-#ifdef HAVE_POLL
-			struct pollfd pollFileDescriptor;
+			/* TODO use resource that only holds the last pointer and frees it on swap */
+			EnsureReleaseResource((MemoryContextCallbackFunction) (&FreeWaitEventSet),
```

Comment:
This call might come too late if there is an exception while constructing the wait event set. We should probably do this at the beginning (and resolve the TODO).